File: master-thesis-config-2/mtc-2-nodejs-algorithm/src/algorithms/road-plan/logic/road.ts
import Coordinate from './coordinate';
import { CoordinateDto } from '../models/coordinateDto';
import roadPlanConfig from '../helpers/roadPlanConfig';

class Road {
  coordinates: Coordinate[];
  distance: number = 0;
  fitnessRatio: number;

  /**
   *  Road Constructor
   */
  constructor(coordinates: Coordinate[]) {
    this.coordinates = coordinates;
    this.distance = this.calculateDistance();
    this.fitnessRatio = this.calculateFitnessRatio();
  }

  rearrange = () => {
    let temp: Coordinate[] = [...this.coordinates];
    let count = temp.length;

    while (count > 1) {
      count--;
      let randomElementIndex = Math.floor(Math.random() * (count - 0 + 1) + 0);
      let randomElement = temp[randomElementIndex];
      temp[randomElementIndex] = temp[count];
      temp[count] = randomElement;
    }

    return new Road(temp);
  };

  calculateDistance = () => {
    let totalDistance = 0;
    for (let i = 0; i < this.coordinates.length; i++) {
      const nextItemIndex = (i + 1) % this.coordinates.length;
      totalDistance += this.coordinates[i].distance(
        this.coordinates[nextItemIndex]
      );
    }

    return totalDistance;
  };

  calculateFitnessRatio = () => {
    if (this.distance === 0) {
      this.distance = this.calculateDistance();
    }

    return 1.0 / this.distance;
  };

  mutation = () => {
    let coords: Coordinate[] = [...this.coordinates];
    let prob = Math.random();
    let road: null | Road = null;

    if (roadPlanConfig.mutationProbability > prob) {
      const swappedIndexOne = Math.floor(
        Math.random() * (this.coordinates.length - 1 - 0 + 1) + 0
      );

      const swappedIndexTwo = Math.floor(
        Math.random() * (this.coordinates.length - 1 - 0 + 1) + 0
      );

      let temp = coords[swappedIndexOne];
      coords[swappedIndexOne] = coords[swappedIndexTwo];
      coords[swappedIndexTwo] = temp;
    }

    road = new Road(coords);

    return road;
  };

  crossing = (road: Road) => {
    const firstPoint = Math.floor(
      Math.random() * (road.coordinates.length - 1 - 0 + 1) + 0
    );

    const secondPoint = Math.floor(
      Math.random() * (road.coordinates.length - 1 - firstPoint + 1) +
        firstPoint
    );

    let returnedRoad: null | Road = null;

    let firstRange = this.coordinates.slice(
      firstPoint,
      firstPoint + (secondPoint - firstPoint + 1)
    );
    let exceptFirstRange = road.coordinates.filter(
      (p) => firstRange.find((x) => x.address === p.address) === undefined
    );
    let merged = [
      ...exceptFirstRange.filter((coord, idx) => idx <= firstPoint),
      ...firstRange,
      ...exceptFirstRange.filter((coord, idx) => idx > firstPoint),
    ];

    returnedRoad = new Road(merged);

    return returnedRoad;
  };
}

export default Road;

File: master-thesis-config-2/mtc-2-dotnet/Domain/Algorithms/Road.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Domain.Algorithms.Models;

namespace Domain.Algorithms
{
    public class Road
    {
        public double Distance { get; set; } = 0.0;
        public double FitnessRatio { get; set; }
        public List<Coordinate> Coordinates { get; set; }
        
        public static Random randomGenerator { get; set; }


        public Road(List<Coordinate> coordinates)
        {
            Coordinates = coordinates;
            prepareRoadParams();
            randomGenerator = new Random();
        }

        private void prepareRoadParams()
        {
            Distance = calculateDistance();
            FitnessRatio = calculateFitnessRatio();
        }

        private double calculateDistance()
        {
            double totalDistance = 0.0;
            for (int i = 0; i < Coordinates.Count; i++)
            {
                var nextItemIndex = (i + 1) % Coordinates.Count;
                totalDistance += Coordinates[i].distance(Coordinates[nextItemIndex]);
            }

            return totalDistance;
        }

        private double calculateFitnessRatio()
        {
            if (Distance == 0)
            {
                Distance = calculateDistance();
            }

            return 1.0 / Distance;
        }
        
        public override string ToString()
        {
            string roadString = String.Empty;
            for (int i = 0; i < Coordinates.Count; i++)
            {
                roadString += $"{Coordinates[i]} -> ";
            }
            roadString += $"{Coordinates[0]}";
            return roadString;
        }

        public Road PerformMutation()
        {
            var coords = new List<Coordinate>(Coordinates);
            var prob = randomGenerator.NextDouble();
            Road road = null;

            if (Config.mutationProbability > prob)
            {
                int swappedIndexOne = randomGenerator.Next(0, Coordinates.Count);
                int swappedIndexTwo = randomGenerator.Next(0, Coordinates.Count);

                var temp = coords[swappedIndexOne];
                coords[swappedIndexOne] = coords[swappedIndexTwo];
                coords[swappedIndexTwo] = temp;
            }
            
            road = new Road(coords);
            
            return road;
        }

        public Road PerformCrossing(Road road)
        {
            int i = randomGenerator.Next(0, road.Coordinates.Count);
            int j = randomGenerator.Next(i, road.Coordinates.Count);
            Road returnedRoad = null;
            
            List<Coordinate> s = Coordinates.GetRange(i, j - i + 1);
            List<Coordinate> ms = road.Coordinates.Except(s).ToList();
            List<Coordinate> c = ms.Take(i)
                .Concat(s)
                .Concat( ms.Skip(i) )
                .ToList();
            
            returnedRoad = new Road(c);
            
            return returnedRoad;
        }

        public Road Rearrange()
        {
            List<Coordinate> tmp = new List<Coordinate>(Coordinates);
            int n = tmp.Count;

            while (n > 1)
            {
                n--;
                var k = randomGenerator.Next( n + 1);
                var v = tmp[k];
                tmp[k] = tmp[n];
                tmp[n] = v;
            }

            return new Road(tmp);
        }
    }
}